File: flutter-app/lib/services/api_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/report.dart';
import 'api_config.dart';

class ApiClient {
  Future<List<Report>> fetchReports() async {
    final uri = Uri.parse('${ApiConfig.baseUrl}/reports');
    final response = await http.get(uri, headers: {'Content-Type': 'application/json'});

    if (response.statusCode < 200 || response.statusCode > 299) {
      throw Exception('Failed to load reports (${response.statusCode})');
    }

    final body = jsonDecode(response.body);
    if (body is! List) {
      throw Exception('Unexpected response format');
    }

    return body.map((item) => Report.fromJson(item as Map<String, dynamic>)).toList();
  }
}

File: ios-app/UniSloveIOS/Networking/APIClient.swift
import Foundation

enum APIError: LocalizedError {
    case invalidURL
    case invalidResponse
    case requestFailed(Int, String)
    case decodingFailed

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid API URL"
        case .invalidResponse:
            return "Invalid server response"
        case .requestFailed(let code, let message):
            return "Request failed (\(code)): \(message)"
        case .decodingFailed:
            return "Failed to decode response data"
        }
    }
}

final class APIClient {
    static let shared = APIClient()
    private init() {}

    func fetchReports() async throws -> [Report] {
        guard let url = URL(string: "\(APIConfig.baseURL)/reports") else {
            throw APIError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if !APIConfig.clerkUserId.isEmpty {
            request.setValue(APIConfig.clerkUserId, forHTTPHeaderField: "x-clerk-user-id")
        }
        if !APIConfig.clerkEmail.isEmpty {
            request.setValue(APIConfig.clerkEmail, forHTTPHeaderField: "x-clerk-email")
        }
        if !APIConfig.clerkUsername.isEmpty {
            request.setValue(APIConfig.clerkUsername, forHTTPHeaderField: "x-clerk-username")
        }
        if !APIConfig.clerkFullName.isEmpty {
            request.setValue(APIConfig.clerkFullName, forHTTPHeaderField: "x-clerk-full-name")
        }

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let http = response as? HTTPURLResponse else {
            throw APIError.invalidResponse
        }

        guard (200 ... 299).contains(http.statusCode) else {
            let message = String(data: data, encoding: .utf8) ?? "Unknown error"
            throw APIError.requestFailed(http.statusCode, message)
        }

        do {
            return try JSONDecoder().decode([Report].self, from: data)
        } catch {
            throw APIError.decodingFailed
        }
    }
}
